core/domain/exp_services.py

```diff
@@ -1039,6 +1039,27 @@ def delete_exploration(committer_id, exploration_id, force_deletion=False):
     delete_state_id_mapping_model_for_exploration(
         exploration_id, exploration_version)
 
+    # Remove from subscribers.
+    taskqueue_services.defer(
+        delete_exploration_from_subscribed_users,
+        taskqueue_services.QUEUE_NAME_EVENTS,
+        exploration_id)
+
+
+def delete_exploration_from_subscribed_users(exploration_id):
+    """Remove exploration from all subscribers' activity_ids.
+
+    Args:
+        exploration_id: The id of the exploration to delete.
+    """
+    subscription_models = user_models.UserSubscriptionsModel.query(
+        user_models.UserSubscriptionsModel.activity_ids.IN(
+            [exploration_id])).fetch()
+
+    for model in subscription_models:
+        model.activity_ids.remove(exploration_id)
+        model.put()
```

Comment:
Perhaps do all the changes and finally a single put_multi()? One call to the datastore, instead of N.